lib/view_component/test_helpers.rb

```diff
@@ -162,5 +173,71 @@ def with_request_url(path)
     def build_controller(klass)
       klass.new.tap { |c| c.request = request }.extend(Rails.application.routes.url_helpers)
     end
+
+    # This is a simpler version of {Capybara::Session}.
```

Comment:
I think this makes sense, but I'm not sure I completely understand the reason not to use the upstream class. Could you explain why we can't, or wouldn't want to utilize `Capybara::Session`?

Reply:
`Capybara::Session` expects a browser, or rather an application or a driver. The `Capybara::Session` class is quite big and complicated. It has quite a few methods that doesn't make sense in the context of testing a view component: window handling, executing JavaScript, handling alerts, navigation, response headers, status code and similar things.

Reply:
Thanks for the clarification.

I'm hesitant about having to copy the internals of Capybara classes to create our own class since it feels like we're relying on internal logic/behavior that could break any time Capybara decides to update the internals of their API's.

I'd really like to see if we could avoid that. I had two thoughts around avoiding including copy/pasted code:

* Could we write a wrapper around `Capybara::Session` that exposes only the public methods we need?
* Could we make a contribution to Capybara that would let us avoid having to copy/paste their code into ViewComponent?

For the wrapper idea, I was thinking something very roughly like:

```
class SimpleSession
  # only delegate the methods we want to expose to ViewComponent tests
  delegate :foo, :bar, :baz, to: :@original_session

  def initialize
    @original_session = Capybara::Session.new
  end
end
```

Reply:
> I'm hesitant about having to copy the internals of Capybara classes to create our own class since it feels like we're relying on internal logic/behavior that could break any time Capybara decides to update the internals of their API's.

I understand your reasoning, but the code is not relying on the internals of Capybara. What the `SimpleSession` is doing is mirroring some of the public API of `Capybara::Session`. It's an implementation detail that I've copied some of the code from Capybara. I've looked through the code in `SimpleSession`, I don't see how Capybara could make non-breaking changes that would break the code in `SimpleSession`. If you have any concrete examples of that, please let me know.

Reply:
> Could we write a wrapper around Capybara::Session that exposes only the public methods we need?

No, not without relying on the internal code of `Capybara::Session`.

> Could we make a contribution to Capybara that would let us avoid having to copy/paste their code into ViewComponent?

I guess we could contribute `SimpleSession` to Capybara if they're interested. I'm actually already using the `SimpleSession` in regular view tests. But it would be a bit more work contributing to Capybara and try to deduplicate the code in the session classes.